clippy_lints/src/static_items_large_align.rs

```diff
@@ -0,0 +1,186 @@
+    ///     println!("{:#x}", &X as *const _ as usize);
+    ///     let b = Box::new(Aligned(0));       // Good
+    ///     println!("{:#x}", Box::into_raw(b) as usize);
+    /// }
+    /// ```
+    #[clippy::version = "1.61.0"]
+    pub STATIC_ITEMS_LARGE_ALIGN,
+    pedantic,
+    "static items with large type alignment, which may trigger unsound problems"
+}
+impl_lint_pass!(StaticItemsLargeAlign => [STATIC_ITEMS_LARGE_ALIGN]);
+
+impl LateLintPass<'_> for StaticItemsLargeAlign {
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if_chain! {
+            if let ItemKind::Static(hir_ty, _, _) = item.kind;
+            let ty = hir_ty_to_ty(cx.tcx, hir_ty);
+            let mut visited_tys = FxHashSet::default();
+            let mut intermediate_tys = Vec::new();
+            if let Some(la_ty) = self.check_ty_alignment(cx.tcx, ty, &mut visited_tys, &mut intermediate_tys);
+            then {
+                let mut span_notes: Vec<(Span, String)> = Vec::new();
+                if !intermediate_tys.is_empty() {
+                    let top_ty = intermediate_tys[0].ty;
+                    if !top_ty.is_adt() {
+                        span_notes.push((
+                            hir_ty.span,
+                            format!("this {} contains an inner type with large alignment", top_ty.prefix_string(cx.tcx)),
+                        ));
+                    }
+                    intermediate_tys.iter()
+                        .filter_map(|im_ty| Self::report_im_ty(cx, im_ty))
+                        .for_each(|ss| span_notes.push(ss));
+                }
+                span_notes.push(self.report_la_ty(cx, &la_ty));
+
+                span_lint_and_then(
+                    cx,
+                    STATIC_ITEMS_LARGE_ALIGN,
+                    item.span,
+                    "this static item (itself or its subfield) has a type alignment,\n\
+                    which is larger than page size and may not be fulfilled,\n\
+                    for more information, see <https://github.com/rust-lang/rust/issues/70022>.",
+                    move |diag| {
+                        for (span, s) in span_notes {
+                            diag.span_note(span, s.as_str());
+                        }
+                    }
+                );
+            }
+        }
+    }
+}
+
+impl StaticItemsLargeAlign {
+    /// It checks a type with the following steps:
+    /// 1. Check if the type is already visited (for a static item),
+    ///     if not, continue;
+    ///     otherwise, return `None` early.
+    /// 2. Push the type in the checked types.
+    /// 3. Pick out this type if its kind is among adt, tuple, array, ref or raw ptr to them;
+    ///     otherwise return `None`.
+    /// 4. Check if its (or its inner fields') alignment are larger than page size.
+    /// 5. Return one of them;
+    ///     otherwise pop the current checked type and return `None`.
+    fn check_ty_alignment<'tcx>(
```

Comment:
Doesn't rustc have a query that just gives you the type alignment? :thinking: Implementing type alignment calculation in Clippy feels a bit wrong to me.